README.md
# Open vSwitch Collection
[![CI](https://zuul-ci.org/gated.svg)](https://dashboard.zuul.ansible.com/t/ansible/project/github.com/ansible-collections/openvswitch.openvswitch) <!--[![Codecov](https://img.shields.io/codecov/c/github/ansible-collections/vyos)](https://codecov.io/gh/ansible-collections/openvswitch.openvswitch)-->

The Open vSwitch collection includes a variety of Ansible content to help automate the management of Open vSwitch.

<!--start requires_ansible-->
## Ansible version compatibility

This collection has been tested against following Ansible versions: **>=2.9.10,<2.11**.

Plugins and modules within a collection may be tested with only specific Ansible versions.
A collection may contain metadata that identifies these versions.
PEP440 is the schema used to describe the versions of Ansible.
<!--end requires_ansible-->

### Supported connections
The Open vSwitch collection supports local connections only.

## Included content

Click the ``Content`` button to see the list of content included in this collection.

<!--start collection content-->
### Modules
Name | Description
--- | ---
[openvswitch.openvswitch.openvswitch_bond](https://github.com/ansible-collections/openvswitch.openvswitch/blob/main/docs/openvswitch.openvswitch.openvswitch_bond_module.rst)|Manage Open vSwitch bonds
[openvswitch.openvswitch.openvswitch_bridge](https://github.com/ansible-collections/openvswitch.openvswitch/blob/main/docs/openvswitch.openvswitch.openvswitch_bridge_module.rst)|Manage Open vSwitch bridges
[openvswitch.openvswitch.openvswitch_db](https://github.com/ansible-collections/openvswitch.openvswitch/blob/main/docs/openvswitch.openvswitch.openvswitch_db_module.rst)|Configure open vswitch database.
[openvswitch.openvswitch.openvswitch_port](https://github.com/ansible-collections/openvswitch.openvswitch/blob/main/docs/openvswitch.openvswitch.openvswitch_port_module.rst)|Manage Open vSwitch ports

<!--end collection content-->

## Installing this collection

You can install the Open vSwitch collection with the Ansible Galaxy CLI:

    ansible-galaxy collection install openvswitch.openvswitch

You can also include it in a `requirements.yml` file and install it with `ansible-galaxy collection install -r requirements.yml`, using the format:

```yaml
---
collections:
  - name: openvswitch.openvswitch
```
## Using this collection

You can call modules by their Fully Qualified Collection Namespace (FQCN), such as `openvswitch.openvswitch.openvswitch_port`.
The following example task replaces configuration changes in the existing configuration on a Open vSwitch network device, using the FQCN:

```yaml
---
  - name: Creates port eth2 on bridge br-ex
    openvswitch.openvswitch.openvswitch_port:
      bridge: br-ex
      port: eth2
    state: present
```

**NOTE**: For Ansible 2.9, you may not see deprecation warnings when you run your playbooks with this collection. Use this documentation to track when a module is deprecated.


### See Also:

* [Ansible Using collections](https://docs.ansible.com/ansible/latest/user_guide/collections_using.html) for more details.

## Contributing to this collection

We welcome community contributions to this collection. If you find problems, please open an issue or create a PR against the [Open vSwitch collection repository](https://github.com/ansible-collections/openvswitch.openvswitch). See [Contributing to Ansible-maintained collections](https://docs.ansible.com/ansible/devel/community/contributing_maintained_collections.html#contributing-maintained-collections) for complete details.

You can also join us on:

- Freenode IRC - ``#ansible-network`` Freenode channel
- Slack - https://ansiblenetwork.slack.com

See the [Ansible Community Guide](https://docs.ansible.com/ansible/latest/community/index.html) for details on contributing to Ansible.

### Code of Conduct
This collection follows the Ansible project's
[Code of Conduct](https://docs.ansible.com/ansible/devel/community/code_of_conduct.html).
Please read and familiarize yourself with this document.


## Changelogs

Release notes are available [here](https://github.com/ansible-collections/openvswitch.openvswitch/blob/main/changelogs/CHANGELOG.rst).

## Roadmap

<!-- Optional. Include the roadmap for this collection, and the proposed release/versioning strategy so users can anticipate the upgrade/update cycle. -->

## More information

- [Ansible network resources](https://docs.ansible.com/ansible/latest/network/getting_started/network_resources.html)
- [Ansible Collection overview](https://github.com/ansible-collections/overview)
- [Ansible User guide](https://docs.ansible.com/ansible/latest/user_guide/index.html)
- [Ansible Developer guide](https://docs.ansible.com/ansible/latest/dev_guide/index.html)
- [Ansible Community code of conduct](https://docs.ansible.com/ansible/latest/community/code_of_conduct.html)

## Licensing

GNU General Public License v3.0 or later.

See [LICENSE](https://www.gnu.org/licenses/gpl-3.0.txt) to see the full text.


CHANGELOG.md
- Added code to perform get operation. Added testcases to test the get operation
- Prepare for v2.0.2 (#82)
- existing vlan is a string, convert wanted vlan before comparing (#81)
- added 'issues' to 'galaxy.yml' (#78)
- Fix pylint issue (#79)
- Update README: freenode -> libera.chat (#76)
- To update the tagged version to 2.0.1 (#71)
- Fix validate-modules failure (#73)
- Move CHANGELOG.rst to top-level (#70)
- Remove tests/sanity/requirements.txt (#69)
- Prepare for release (#68)
- Allow setting multiple properties on a port (#66)
- Allow deleting key from db without specifying value (#65)
- Add changelog for 1.1.0 (#62)
- Update docs for for release (#61)
- Adds Open vSwitch bond management module (#58)
- Fix galaxy.yaml (#60)
- Remove ansible from requirements.txt (#57)
- Prepare for 1.0.5 release (#53)
- Add changelog for 1.0.3 (#51)
- Add changelog (#50)
- Update docs (#48)
- Remove ignores for sanity test (#47)
- Add missing dependencies for unit testing (#46)
- Update references to main (#44)
- Add changelog files (#41)
- Updates for release readiness (#39)
- Fix error when '=' is used in the dpdk-extra (#34)
- added contributor link etc (#37)
- Update collection version in README (#35)
- Update runtime.yml requires_ansible version (#33)
- Updates needed to pass ansible-test sanity (#27)
- Add bindep.txt support (#23)
- Add runtime.yml file (#21)
- Add coverage support to ovs  (#17)
- Merge pull request #20 from danielmellado/update_module_docs
- Add OVS modules DOC
- Merge pull request #19 from danielmellado/adapt_ovs_collection
- Remove ANSIBLE_METADATA
- Adapt examples to collection format
- Merge pull request #18 from samccann/ovs_reame
- update ovs readme
- Merge pull request #16 from JPrause/openvswtich_tag
- Adding labels to the tags section
- Merge pull request #13 from pabelanger/temp/ansible.cfg
- Merge pull request #12 from NilashishC/master
- Add missing ansible.cfg for integration tests
- Add pytest as a requirement
- Merge pull request #10 from pabelanger/master
- Update repo location in galaxy.yml
- Merge pull request #6 from pabelanger/master
- Move python dependencies to top level
- Merge pull request #5 from pabelanger/master
- Add LICENSE file
- Merge pull request #3 from pabelanger/master
- Add prepare_ovs_tests files
- Merge pull request #2 from CaptTrews/master
- Updated from network content collector
- Merge pull request #1 from pabelanger/master
- Bootstrap repo for testing
- Initial commit


tests/unit/mock/path.py
from ansible_collections.openvswitch.openvswitch.tests.unit.compat.mock import (
    MagicMock,
)
from ansible.utils.path import unfrackpath


mock_unfrackpath_noop = MagicMock(
    spec_set=unfrackpath, side_effect=lambda x, *args, **kwargs: x
)


tests/unit/mock/loader.py
# (c) 2012-2014, Michael DeHaan <michael.dehaan@gmail.com>
#
# This file is part of Ansible
#
# Ansible is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# Ansible is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.

# Make coding more python3-ish
from __future__ import absolute_import, division, print_function

__metaclass__ = type

import os

from ansible.errors import AnsibleParserError
from ansible.parsing.dataloader import DataLoader
from ansible.module_utils._text import to_bytes, to_text


class DictDataLoader(DataLoader):
    def __init__(self, file_mapping=None):
        file_mapping = {} if file_mapping is None else file_mapping
        assert type(file_mapping) == dict

        super(DictDataLoader, self).__init__()

        self._file_mapping = file_mapping
        self._build_known_directories()
        self._vault_secrets = None

    def load_from_file(self, path, cache=True, unsafe=False):
        path = to_text(path)
        if path in self._file_mapping:
            return self.load(self._file_mapping[path], path)
        return None

    # TODO: the real _get_file_contents returns a bytestring, so we actually convert the
    #       unicode/text it's created with to utf-8
    def _get_file_contents(self, file_name):
        if file_name in self._file_mapping:
            return (to_bytes(self._file_mapping[file_name]), False)
        else:
            raise AnsibleParserError("file not found: %s" % file_name)

    def path_exists(self, path):
        path = to_text(path)
        return path in self._file_mapping or path in self._known_directories

    def is_file(self, path):
        path = to_text(path)
        return path in self._file_mapping

    def is_directory(self, path):
        path = to_text(path)
        return path in self._known_directories

    def list_directory(self, path):
        ret = []
        path = to_text(path)
        for x in list(self._file_mapping.keys()) + self._known_directories:
            if x.startswith(path):
                if os.path.dirname(x) == path:
                    ret.append(os.path.basename(x))
        return ret

    def is_executable(self, path):
        # FIXME: figure out a way to make paths return true for this
        return False

    def _add_known_directory(self, directory):
        if directory not in self._known_directories:
            self._known_directories.append(directory)

    def _build_known_directories(self):
        self._known_directories = []
        for path in self._file_mapping:
            dirname = os.path.dirname(path)
            while dirname not in ("/", ""):
                self._add_known_directory(dirname)
                dirname = os.path.dirname(dirname)

    def push(self, path, content):
        rebuild_dirs = False
        if path not in self._file_mapping:
            rebuild_dirs = True

        self._file_mapping[path] = content

        if rebuild_dirs:
            self._build_known_directories()

    def pop(self, path):
        if path in self._file_mapping:
            del self._file_mapping[path]
            self._build_known_directories()

    def clear(self):
        self._file_mapping = dict()
        self._known_directories = []

    def get_basedir(self):
        return os.getcwd()

    def set_vault_secrets(self, vault_secrets):
        self._vault_secrets = vault_secrets


plugins/modules/openvswitch_db.py
#!/usr/bin/python

#
# (c) 2015, Mark Hamilton <mhamilton@vmware.com>
# Portions copyright @ 2015 VMware, Inc.
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

from __future__ import absolute_import, division, print_function

__metaclass__ = type


DOCUMENTATION = """
module: openvswitch_db
author: Mark Hamilton (@markleehamilton) <mhamilton@vmware.com>
short_description: Configure open vswitch database.
requirements:
- ovs-vsctl >= 2.3.3
description:
- Set column values in record in database table.
version_added: 1.0.0
options:
  state:
    required: false
    description:
    - Configures the state of the key. When set to I(present), the I(key) and I(value)
      pair will be set on the I(record) and when set to I(absent) the I(key) will
      not be set.
    default: present
    choices:
    - present
    - absent
    type: str
  table:
    required: true
    description:
    - Identifies the table in the database.
    type: str
  record:
    required: true
    description:
    - Identifies the record in the table.
    type: str
  col:
    required: true
    description:
    - Identifies the column in the record.
    type: str
  key:
    required: false
    description:
    - Identifies the key in the record column, when the column is a map type.
    type: str
  value:
    description:
    - Expected value for the table, record, column and key.
    type: str
  timeout:
    required: false
    default: 5
    description:
    - How long to wait for ovs-vswitchd to respond
    type: int
  database_socket:
    description:
    - Path/ip to datbase socket to use
    - Default path is used if not specified
    - Path should start with 'unix:' prefix
    type: str
"""

EXAMPLES = """
# Increase the maximum idle time to 50 seconds before pruning unused kernel
# rules.
- openvswitch.openvswitch.openvswitch_db:
    table: open_vswitch
    record: .
    col: other_config
    key: max-idle
    value: 50000

# Disable in band copy
- openvswitch.openvswitch.openvswitch_db:
    table: Bridge
    record: br-int
    col: other_config
    key: disable-in-band
    value: true

# Remove in band key
- openvswitch.openvswitch.openvswitch_db:
    state: present
    table: Bridge
    record: br-int
    col: other_config
    key: disable-in-band

# Mark port with tag 10
- openvswitch.openvswitch.openvswitch_db:
    table: Port
    record: port0
    col: tag
    value: 10

# Mark port with tag 10 for OVSDB with socket in /opt/second.sock
- openvswitch.openvswitch.openvswitch_db:
    table: Port
    record: port0
    col: tag
    value: 10
    database_socket: unix:/opt/second.sock
"""
import re

from ansible.module_utils.basic import AnsibleModule

# Regular expression for map type, must not be empty
NON_EMPTY_MAP_RE = re.compile(r"{.+}")
# Regular expression for a map column type
MAP_RE = re.compile(r"{.*}")


def map_obj_to_commands(want, have, module):
    """ Define ovs-vsctl command to meet desired state """
    commands = list()

    if module.params["state"] == "absent":
        if "key" in have.keys():
            templatized_command = (
                "%(ovs-vsctl)s -t %(timeout)s remove %(table)s %(record)s "
                "%(col)s %(key)s"
            )
            if module.params.get("value"):
                templatized_command += "=%(value)s"
            commands.append(templatized_command % module.params)
        elif module.params["key"] is None:
            templatized_command = (
                "%(ovs-vsctl)s -t %(timeout)s remove %(table)s %(record)s "
                "%(col)s"
            )
            commands.append(templatized_command % module.params)
    else:
        if want == have:
            # Nothing to commit
            return commands
        if module.params["key"] is None:
            templatized_command = (
                "%(ovs-vsctl)s -t %(timeout)s set %(table)s %(record)s "
                "%(col)s=%(value)s"
            )
            commands.append(templatized_command % module.params)
        else:
            templatized_command = (
                "%(ovs-vsctl)s -t %(timeout)s set %(table)s %(record)s "
                "%(col)s:%(key)s=%(value)s"
            )
            commands.append(templatized_command % module.params)

    return commands


def map_config_to_obj(module):
    templatized_command = (
        "%(ovs-vsctl)s -t %(timeout)s list %(table)s %(record)s"
    )
    command = templatized_command % module.params
    rc, out, err = module.run_command(command, check_rc=True)
    if rc != 0:
        module.fail_json(msg=err)

    match = re.search(
        r"^" + module.params["col"] + r"(\s+):(\s+)(.*)$", out, re.M
    )

    col_value = match.group(3)

    # Map types require key argument
    has_key = module.params["key"] is not None
    is_map = MAP_RE.match(col_value)
    if is_map and not has_key:
        module.fail_json(
            msg="missing required arguments: key for map type of column"
        )

    col_value_to_dict = {}
    if NON_EMPTY_MAP_RE.match(col_value):
        for kv in col_value[1:-1].split(", "):
            k, v = kv.split("=", 1)
            col_value_to_dict[k.strip()] = v.strip('"')

    obj = {
        "table": module.params["table"],
        "record": module.params["record"],
        "col": module.params["col"],
    }

    if has_key and is_map:
        if module.params["key"] in col_value_to_dict:
            obj["key"] = module.params["key"]
            obj["value"] = col_value_to_dict[module.params["key"]]
    else:
        obj["value"] = str(col_value.strip())

    return obj


def map_params_to_obj(module):
    if module.params["value"] in ["True", "False"]:
        module.params["value"] = module.params["value"].lower()
    obj = {
        "table": module.params["table"],
        "record": module.params["record"],
        "col": module.params["col"],
        "value": module.params["value"],
    }

    key = module.params["key"]
    if key is not None:
        obj["key"] = key

    return obj


def main():
    """ Entry point for ansible module. """
    argument_spec = {
        "state": {"default": "present", "choices": ["present", "absent"]},
        "table": {"required": True},
        "record": {"required": True},
        "col": {"required": True},
        "key": {"required": False, "no_log": False},
        "value": {"type": "str"},
        "timeout": {"default": 5, "type": "int"},
        "database_socket": {"default": None},
    }

    module = AnsibleModule(
        argument_spec=argument_spec, supports_check_mode=True
    )

    result = {"changed": False}

    # We add ovs-vsctl to module_params to later build up templatized commands
    module.params["ovs-vsctl"] = module.get_bin_path("ovs-vsctl", True)
    if module.params.get("database_socket"):
        module.params["ovs-vsctl"] += " --db=" + module.params.get(
            "database_socket"
        )

    if module.params["state"] == "present" and not module.params["value"]:
        module.fail_json(
            msg="missing required argument value for state: present"
        )

    want = map_params_to_obj(module)
    have = map_config_to_obj(module)

    commands = map_obj_to_commands(want, have, module)
    result["commands"] = commands

    if commands:
        if not module.check_mode:
            for c in commands:
                module.run_command(c, check_rc=True)
        result["changed"] = True

    module.exit_json(**result)


if __name__ == "__main__":
    main()


plugins/modules/openvswitch_port.py
#!/usr/bin/python

# (c) 2013, David Stygstra <david.stygstra@gmail.com>
# Portions copyright @ 2015 VMware, Inc.
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)

from __future__ import absolute_import, division, print_function

__metaclass__ = type


DOCUMENTATION = """
module: openvswitch_port
author: David Stygstra (@stygstra)
short_description: Manage Open vSwitch ports
requirements:
- ovs-vsctl
description:
- Manage Open vSwitch ports
version_added: 1.0.0
options:
  bridge:
    required: true
    description:
    - Name of bridge to manage
    type: str
  port:
    required: true
    description:
    - Name of port to manage on the bridge
    type: str
  tag:
    description:
    - VLAN tag for this port. Must be a value between 0 and 4095.
    type: str
  state:
    default: present
    choices:
    - present
    - absent
    description:
    - Whether the port should exist
    type: str
  timeout:
    default: 5
    description:
    - How long to wait for ovs-vswitchd to respond
    type: int
  external_ids:
    default: {}
    description:
    - Dictionary of external_ids applied to a port.
    type: dict
  set:
    description:
    - Set multiple properties on a port.
    type: list
    elements: str
  database_socket:
    description:
    - Path/ip to datbase socket to use
    - Default path is used if not specified
    - Path should start with 'unix:' prefix
    type: str
"""

EXAMPLES = """
# Creates port eth2 on bridge br-ex
- openvswitch.openvswitch.openvswitch_port:
    bridge: br-ex
    port: eth2
    state: present

# Creates port eth6
- openvswitch.openvswitch.openvswitch_port:
    bridge: bridge-loop
    port: eth6
    state: present
    set: Interface eth6

# Creates port vlan10 with tag 10 on bridge br-ex
- openvswitch.openvswitch.openvswitch_port:
    bridge: br-ex
    port: vlan10
    tag: 10
    state: present
    set: Interface vlan10

# Assign interface id server1-vifeth6 and mac address 00:00:5E:00:53:23
# to port vifeth6 and setup port to be managed by a controller.
- openvswitch.openvswitch.openvswitch_port:
    bridge: br-int
    port: vifeth6
    state: present
  args:
    external_ids:
      iface-id: '{{ inventory_hostname }}-vifeth6'
      attached-mac: 00:00:5E:00:53:23
      vm-id: '{{ inventory_hostname }}'
      iface-status: active

# Plugs port veth0 into brdige br0 for database for OVSDB instance
# with socket unix:/opt/second_ovsdb.sock
- openvswitch.openvswitch.openvswitch_port:
    bridge: br0
    port: veth0
    state: present
    database_socket: unix:/opt/second_ovsdb.sock

"""

from ansible.module_utils.basic import AnsibleModule
from ansible.module_utils.six import iteritems


def _external_ids_to_dict(text):
    text = text.strip()

    if text == "{}":
        return None
    else:
        d = {}

        for kv in text[1:-1].split(","):
            kv = kv.strip()
            k, v = kv.split("=")
            d[k] = v

        return d


def _tag_to_str(text):
    text = text.strip()

    if text == "[]":
        return None
    else:
        return text


def map_obj_to_commands(want, have, module):
    commands = list()

    if module.params["state"] == "absent":
        if have:
            templatized_command = (
                "%(ovs-vsctl)s -t %(timeout)s del-port %(bridge)s %(port)s"
            )
            command = templatized_command % module.params
            commands.append(command)
    else:
        if have:
            if want["tag"] != have["tag"]:
                templatized_command = (
                    "%(ovs-vsctl)s -t %(timeout)s"
                    " set port %(port)s tag=%(tag)s"
                )
                command = templatized_command % module.params
                commands.append(command)

            if want["external_ids"] != have["external_ids"]:
                for k, v in iteritems(want["external_ids"]):
                    if (
                        not have["external_ids"]
                        or k not in have["external_ids"]
                        or want["external_ids"][k] != have["external_ids"][k]
                    ):
                        if v is None:
                            templatized_command = (
                                "%(ovs-vsctl)s -t %(timeout)s"
                                " remove port %(port)s"
                                " external_ids " + k
                            )
                            command = templatized_command % module.params
                            commands.append(command)
                        else:
                            templatized_command = (
                                "%(ovs-vsctl)s -t %(timeout)s"
                                " set port %(port)s"
                                " external_ids:"
                            )
                            command = templatized_command % module.params
                            command += k + "=" + v
                            commands.append(command)
        else:
            templatized_command = (
                "%(ovs-vsctl)s -t %(timeout)s add-port %(bridge)s %(port)s"
            )
            command = templatized_command % module.params

            if want["tag"]:
                templatized_command = " tag=%(tag)s"
                command += templatized_command % module.params

            if want["set"]:
                set_command = ""
                for x in want["set"]:
                    set_command += " -- set {0}".format(x)
                command += set_command

            commands.append(command)

            if want["external_ids"]:
                for k, v in iteritems(want["external_ids"]):
                    templatized_command = (
                        "%(ovs-vsctl)s -t %(timeout)s"
                        " set port %(port)s external_ids:"
                    )
                    command = templatized_command % module.params
                    command += k + "=" + v
                    commands.append(command)

    return commands


def map_config_to_obj(module):
    templatized_command = "%(ovs-vsctl)s -t %(timeout)s list-ports %(bridge)s"
    command = templatized_command % module.params
    rc, out, err = module.run_command(command, check_rc=True)
    if rc != 0:
        module.fail_json(msg=err)

    obj = {}

    if module.params["port"] in out.splitlines():
        obj["bridge"] = module.params["bridge"]
        obj["port"] = module.params["port"]

        templatized_command = (
            "%(ovs-vsctl)s -t %(timeout)s get Port %(port)s tag"
        )
        command = templatized_command % module.params
        rc, out, err = module.run_command(command, check_rc=True)
        obj["tag"] = _tag_to_str(out)

        templatized_command = (
            "%(ovs-vsctl)s -t %(timeout)s get Port %(port)s external_ids"
        )
        command = templatized_command % module.params
        rc, out, err = module.run_command(command, check_rc=True)
        obj["external_ids"] = _external_ids_to_dict(out)

    return obj


def map_params_to_obj(module):
    obj = {
        "bridge": module.params["bridge"],
        "port": module.params["port"],
        "tag": module.params["tag"],
        "external_ids": module.params["external_ids"],
        "set": module.params["set"],
    }

    return obj


def main():
    """ Entry point. """
    argument_spec = {
        "bridge": {"required": True},
        "port": {"required": True},
        "state": {"default": "present", "choices": ["present", "absent"]},
        "timeout": {"default": 5, "type": "int"},
        "external_ids": {"default": None, "type": "dict"},
        "tag": {"default": None},
        "database_socket": {"default": None},
        "set": {"required": False, "type": "list", "elements": "str"},
    }

    module = AnsibleModule(
        argument_spec=argument_spec, supports_check_mode=True
    )

    result = {"changed": False}

    # We add ovs-vsctl to module_params to later build up templatized commands
    module.params["ovs-vsctl"] = module.get_bin_path("ovs-vsctl", True)
    if module.params.get("database_socket"):
        module.params["ovs-vsctl"] += ' --db=' + module.params.get("database_socket")

    want = map_params_to_obj(module)
    have = map_config_to_obj(module)

    commands = map_obj_to_commands(want, have, module)
    result["commands"] = commands

    if commands:
        if not module.check_mode:
            for c in commands:
                module.run_command(c, check_rc=True)
        result["changed"] = True

    module.exit_json(**result)


if __name__ == "__main__":
    main()
